### PR TITLE
[RW-10366][risk=low] Add conditional logic for identity module completion

### PR DESCRIPTION
**Pull request by evrii**
This PR is only to ensure that the data access requirements (DAR) page shows the "Verify your identity with Login.gov" module shows as completed for either ID.me or Login.gov.

In order to test this process, you need a user who has not completed identity verification. I found this process easiest to test by bypassing/un-bypassing the modules within the user admin page.

1. Open the DAR page, the "Verify your identity with Login.gov" module should show as incomplete.
2. Complete or bypass login.gov
3. Refresh the DAR page, and the "Verify your identity with Login.gov" module should now be complete.
4. Revert the login.gov status, and repeat the process for ID.me.

https://github.com/all-of-us/workbench/assets/24514630/1cb74934-a55e-44e0-85f4-423885d8e132

---
**PR checklist**

- [x] I have included an issue ID or "no ticket" in the PR title as outlined in [CONTRIBUTING.md](https://github.com/all-of-us/workbench/blob/main/.github/CONTRIBUTING.md).
- [x] I have included a risk tag of the form `[risk=no|low|moderate|severe]` in the PR title as outlined in [CONTRIBUTING.md](https://github.com/all-of-us/workbench/blob/main/.github/CONTRIBUTING.md).
- [x] I have manually tested this change and my testing process is described above.
- [x] This change includes appropriate automated tests, and I have documented any behavior that cannot be tested with code.
- [x] I have added explanatory comments where the logic is not obvious.
- One or more of the following is true:
  - [ ] This change is intended to complete a JIRA story, so I have checked that all AC are met for that story.
  - [ ] This change fixes a bug, so I have ensured the steps to reproduce are in the Jira ticket or provided above.
  - [ ] This change impacts deployment safety (e.g. removing/altering APIs which are in use), so I have documented the impacts in the description.
  - [ ] This change includes a new feature flag, so I have created and linked new JIRA tickets to (a) turn on the feature flag and (b) remove it later.
  - [x] This change modifies the UI, so I have taken screenshots or recordings of the new behavior and notified the PO and UX designer in [Slack](https://pmi-engteam.slack.com/archives/C02MWP2RN5P).
  - [ ] This change modifies API behavior, so I have run the relevant E2E tests locally because API changes are not covered by our PR checks.
  - [ ] None of the above apply to this change.

### CODE REVIEW

File: ui/src/app/utils/access-utils.tsx [hunk 1]
```diff
@@ -186,6 +190,27 @@ export const getAccessModuleConfig = (
       AccessModule.RASLINKIDME,
       () => ({
         ...apiConfig,
+        isEnabledInEnvironment:
```
Comment (evrii) [inline ui/src/app/utils/access-utils.tsx:193]: While I was developing, I changed the text in here, so I could ensure that the logic for getInitialRTModules was working correctly.

Since the design for this component is going to be done in a separate PR, I made this component equivalent to the current Login.gov component for now, since the most recent change will result in the ID.me component being shown.

File: ui/src/app/utils/access-utils.tsx [hunk 2]
```diff
@@ -314,6 +339,27 @@ export const getAccessModuleConfig = (
   );
 };
 
+export const getInitialRTModules = (): AccessModule[] => {
```
Comment (evrii) [inline ui/src/app/utils/access-utils.tsx:342]: Moved these from static lists to functions in order to make use of feature flagging.

File: api/src/main/java/org/pmiops/workbench/access/AccessUtils.java [hunk 3]
```diff
@@ -26,14 +26,13 @@ public static DbAccessTier getAccessTierByShortNameOrThrow(
             () -> new NotFoundException("Access tier " + accessTierShortName + "not found"));
   }
 
-  // missing ERA_COMMONS (special-cased)
+  // missing ERA_COMMONS, RAS_ID_ME, and RAS_LOGIN_GOV(special-cased)
```
Comment (evrii) [inline api/src/main/java/org/pmiops/workbench/access/AccessUtils.java:29]: Login.gov is no longer strictly required, because users can use ID.me.

File: ui/src/app/utils/access-utils.tsx [hunk 4]
```diff
@@ -459,6 +547,20 @@ export const getAccessModuleStatusByName = (
   );
 };
 
+// Gets status of moduleName, but if it is an identity module,
+// it will return the status of the "most complete" identity module.
+export const getRelativeAccessModuleStatus = (
```
Comment (evrii) [inline ui/src/app/utils/access-utils.tsx:552]: If someone can think of a better name for this function, let me know.

This function exists, because sometimes we need to look up the exact status of a module, but in other cases (like the data access requirements page), we need to. know the status of a group of modules (login.gov and id.me). This function handles the second case.

**Review by Peter-Lavigne — APPROVED** [review]
I have a lot of comments, but I think it's fine overall and works well.

The access module logic is going to get pretty complex - It may still be worth considering using a single access module, instead of two, with special-case logic for how it's completed and that data is stored.

File: ui/src/app/utils/access-utils.tsx [hunk 5]
```diff
@@ -449,14 +500,53 @@ export const getAccessModuleStatusByNameOrEmpty = (
   );
 };
 
+const getAccessModuleStatusForIdentityVerification = (
```
[the remark this answers is not in the document]
Comment (Peter-Lavigne) [reply ui/src/app/utils/access-utils.tsx:503]: (I think my code fails to handle the completion > bypass logic exactly; proceed with caution)

File: ui/src/app/utils/access-utils.tsx (hunk 2 above)
Comment (evrii) [reply ui/src/app/utils/access-utils.tsx:342]: `getInitialRTModules` is strictly used for display logic for the data access requirements page. 

This function helps determines which cards are shown. The DARTitleComponent of Login.gov shows the current identity verification card, but the DARTitleComponent of RASLINKIDME shows the new design where both identity providers are represented. We would not want both of these cards to shown on the same screen, because they are redundant. That is why the Login.gov card is not shown when ID.me is active.

As I am writing this though, I am wondering if it would be more intuitive to always show RASLINKLOGINGOV (never show RASLINKIDME), and make the DARTitleComponent conditionally render. Show the old design when ID.me is not active and show the new design when ID.me is enabled.

Does one approach make more sense than the other to you @Peter-Lavigne?

**Comment by Peter-Lavigne** [comment]
Also, the demo for this PR was done really well. Nice job!

File: ui/src/testing/default-server-config.ts [hunk 6]
```diff
@@ -11,11 +11,17 @@ const defaultAccessModuleConfig: AccessModuleConfig[] = [
...
+    requiredForRTAccess: false,
+    requiredForCTAccess: false,
+  },
   {
     name: AccessModule.RASLINKLOGINGOV,
     expirable: false,
-    requiredForRTAccess: true,
-    requiredForCTAccess: true,
+    requiredForRTAccess: false,
+    requiredForCTAccess: false,
```
[the remark this answers is not in the document]
Comment (evrii) [reply ui/src/testing/default-server-config.ts:24]: I agree, but I am trying to think how to best change this, and whether we need to do.

As far as I can tell, these two fields are used for calculating days until a module expires (not relevant for identity providers) and whether or not to show the RT and CT badges on the Access modules table on the user admin page. I am wondering what we want to show here, because neither of them are strictly required on their own, but it is an either or situation. Is no badge ok?

I am up for suggestions.

image

File: ui/src/app/utils/access-utils.tsx (hunk 2 above)
Comment (Peter-Lavigne) [reply ui/src/app/utils/access-utils.tsx:342]: > is strictly used for display logic for the data access requirements page

Maybe the function should be renamed to indicate this?

> it would be more intuitive to always show RASLINKLOGINGOV (never show RASLINKIDME), and make the DARTitleComponent conditionally render

I think that would be clearer since that'll be the end result once the `enableRasIdMeLinking` feature flag is removed (if this isn't a FF, please let me know).

Maybe there's a way to refactor this to return single concept representing both access modules (like a list containing both) so that it's clear both access modules are represented by one downstream component.

File: ui/src/testing/default-server-config.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (Peter-Lavigne) [reply ui/src/testing/default-server-config.ts:24]: Working backward from what we want the UI to do makes sense.

The badges should be updated in some way. Personally, I haven't used the badges to make any decisions so I'd be OK with removing them. However, I've only ever bypassed test users. It's worth checking with the team if that's helpful for real-user bypasses and what they'd prefer in that field.

We could return "days until a module expires" instead if the badges aren't needed.

File: ui/src/app/utils/access-utils.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (evrii) [reply ui/src/app/utils/access-utils.tsx]: Thank you @Peter-Lavigne ! This is wonderful!

Purpose: Good point! Bypass after a completion is just as valid and drastically simplifies the logic.

Naming: GetIdentityVerification sounds good to me!

Complexity: This is great! I tweaked your suggestion just a bit, but either way, this is much better than what I had initially.

File: ui/src/app/utils/access-utils.tsx (hunk 2 above)
Comment (evrii) [reply ui/src/app/utils/access-utils.tsx]: Switching the component rendering logic to a single module made the code simpler, because it means that the feature flag is not needed when deciding which modules to show. This means that I could switch the module lists to lists instead of functions that return lists.

File: ui/src/app/utils/access-utils.spec.tsx [hunk 7]
```diff
@@ -738,7 +738,7 @@ describe('getAccessModuleStatusByName', () => {
       100,
     ],
     [
-      AccessModule.RASLINKLOGINGOV,
+      AccessModule.RASLINKIDME,
       AccessModule.RASLINKIDME,
```
Comment (Qi77Qi) [inline ui/src/app/utils/access-utils.spec.tsx:742]: it seems duplicates the next line?

Comment (evrii) [reply ui/src/app/utils/access-utils.spec.tsx:742]: The first is the inputted access module and the second is the expected output access module. Those could be the same value or different depending on whether ID.me or Login.gov was completed most recently. 

I am in the process of rewriting these tests to make them more readable.
